### PR TITLE
GH-276 Implement spring cloud messaging annotation  

### PR DESCRIPTION
**Pull request by warrenzhu25**
You can try to go through this one by one commit. Keep this in one big PR will make reviewer easier to get the big picture of whole change.

### CODE REVIEW

**Comment by xscript** [comment]
Per our offline sync, let's avoid such big PRs in the future, because it would take more time for people to understand and provide comment.
So stay tuned for reviews. I am working in progress. 😉

**Comment by Incarnation-p-lee** [comment]
@xscript 
One WIP application from github will block you merge when there is keyword WIP in PR titile. Also need approve under organization.

File: spring-cloud-azure-messaging/src/main/java/com/microsoft/azure/spring/messaging/annotation/AzureListener.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+ *
+ * @author Warren Zhu
+ * @see EnableAzureMessaging
+ * @see AzureListenerAnnotationBeanPostProcessor
+ */
+@Target({ElementType.METHOD, ElementType.ANNOTATION_TYPE})
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@MessageMapping
+public @interface AzureListener {
```
Comment (xscript) [inline spring-cloud-azure-messaging/src/main/java/com/microsoft/azure/spring/messaging/annotation/AzureListener.java]: Is `AzureMessageListener` more clearer to user? `AzureListener` sounds too general.

File: spring-cloud-azure-messaging/src/main/java/com/microsoft/azure/spring/messaging/container/AbstractAzureListenerContainerFactory.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+
+/**
+ * Base {@link ListenerContainerFactory} for Spring's base container implementation.
+ *
+ * @param <C> the container type
+ * @author Warren Zhu
+ * @see AbstractAzureListenerEndpoint
+ */
+
```
Comment (xscript) [inline spring-cloud-azure-messaging/src/main/java/com/microsoft/azure/spring/messaging/container/AbstractAzureListenerContainerFactory.java]: Let's unify spacing in all files.

File: spring-cloud-azure-messaging/src/main/java/com/microsoft/azure/spring/messaging/endpoint/AbstractAzureListenerEndpoint.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+     */
+    protected abstract AzureMessageHandler createMessageHandler(MessageListenerContainer container);
+
+    /**
+     * Return a description for this endpoint.
+     * <p>Available to subclasses, for inclusion in their {@code toString()} result.
+     */
+    protected StringBuilder getEndpointDescription() {
+        StringBuilder result = new StringBuilder();
+        return result.append(getClass().getSimpleName()).append("[").append(this.id).append("] destination=").
```
Comment (xscript) [inline spring-cloud-azure-messaging/src/main/java/com/microsoft/azure/spring/messaging/endpoint/AbstractAzureListenerEndpoint.java]: `String.format` is more clearer?

File: spring-cloud-azure-messaging/src/main/java/com/microsoft/azure/spring/messaging/container/AbstractAzureListenerContainerFactory.java (hunk 2 above)
Comment (warrenzhu25) [reply spring-cloud-azure-messaging/src/main/java/com/microsoft/azure/spring/messaging/container/AbstractAzureListenerContainerFactory.java]: I suppose IDE should have this kind of format. Let me see what's wrong with this.

File: spring-cloud-azure-samples/pom.xml [hunk 4]
```diff
@@ -29,6 +29,7 @@
         <module>spring-cloud-azure-servicebus-topic-binder-sample</module>
         <module>spring-cloud-azure-storage-queue-operation-sample</module>
         <module>spring-cloud-azure-storage-queue-integration-sample</module>
+        <module>azure-messaging-sample</module>
```
Comment (xscript) [inline spring-cloud-azure-samples/pom.xml]: nit: The package is already named as `spring-cloud-azure-messaging`. Maybe we can still use `spring-cloud-azure-messaging-sample` to keep consistent with other samples.

Comment (warrenzhu25) [reply spring-cloud-azure-samples/pom.xml]: I have plan to remove the long redundant spring-cloud prefix for all submodules, please refer #308

File: spring-cloud-azure-messaging/src/main/java/com/microsoft/azure/spring/messaging/listener/DefaultAzureMessageHandler.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+     * processing an incoming {@link Message}.
+     */
+    public void setHandlerMethod(InvocableHandlerMethod handlerMethod) {
+        this.handlerMethod = handlerMethod;
+        this.messagePayloadType = resolveMessagePayloadType(this.handlerMethod);
+    }
+
+    private String createMessagingErrorMessage(String description) {
+        InvocableHandlerMethod handlerMethod = getHandlerMethod();
+        StringBuilder sb =
```
Comment (xscript) [inline spring-cloud-azure-messaging/src/main/java/com/microsoft/azure/spring/messaging/listener/DefaultAzureMessageHandler.java]: nit: `Spring.format` might be a better choice.

**Comment by xscript** [comment]
After renaming `AzureListener` to `AzureMessageListener`, we will have to rename related classes for consistency?
A brief list for your reference.
- `AzureListenerAnnotationBeanPostProcessor`
- `AzureListenerConfigurer`
- `AbstractAzureListenerContainerFactory`
- `DefaultAzureListenerContainerFactory`
- `AbstractAzureListenerEndpoint`
- `AzureListenerEndpoint`
- `AzureListenerEndpointRegistrar`
- `AzureListenerEndpointRegistry`
- `MethodAzureListenerEndpoint`

File: spring-cloud-azure-messaging/src/main/java/com/microsoft/azure/spring/messaging/listener/AzureMessageHandler.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+ * Licensed under the MIT License. See LICENSE in the project root for
+ * license information.
+ */
+
+package com.microsoft.azure.spring.messaging.listener;
+
+import org.springframework.lang.Nullable;
+import org.springframework.messaging.MessageHandler;
+
+public interface AzureMessageHandler extends MessageHandler {
```
[the remark this answers is not in the document]
Comment (warrenzhu25) [reply spring-cloud-azure-messaging/src/main/java/com/microsoft/azure/spring/messaging/listener/AzureMessageHandler.java]: Good point. I'll rename another into `AzureIntegrationMessageHanlder` in another PR.

**Pull request merged**